fendermint/eth/api/src/mpool.rs

```diff
@@ -25,22 +27,81 @@ const RETRY_SLEEP_SECS: u64 = 5;
 
-pub fn start_tx_cache_clearing(client: FendermintClient<HybridClient>, cache: TransactionCache) {
+/// Buffer out-of-order messages until they can be sent to the chain.
+#[derive(Clone)]
+pub struct TransactionBuffer(pub Cache<Address, BTreeMap<Nonce, ChainMessage>>);
+
+impl TransactionBuffer {
+    /// Insert a transaction we could not submit straight away into the buffer.
+    pub fn insert(&self, sender: Address, nonce: Nonce, msg: ChainMessage) {
+        self.0.with(|c| {
+            let buffer = c.entry(sender).or_insert_with(BTreeMap::new);
+            // Overwrite any previous entry to protect against DoS attack; it wouldn't make sense to submit them anyway.
+            buffer.insert(nonce, msg);
+        })
+    }
+
+    /// Remove all (sender, nonce) pairs which were included in a block.
+    fn remove_many<'a, I>(&self, txs: I)
+    where
+        I: Iterator<Item = (&'a Address, Nonce)>,
+    {
+        self.0.with(|c| {
+            for (sender, nonce) in txs {
+                if let Some(buffer) = c.get_mut(sender) {
+                    buffer.remove(&nonce);
+                }
+            }
+        })
+    }
+
+    /// Gather any messages that have been enabled by transactions added to a block.
+    ///
+    /// These are removed from the cache, submission is only attempted once.
+    fn remove_enabled<'a, I>(&self, txs: I) -> Vec<(Address, Nonce, ChainMessage)>
+    where
+        I: Iterator<Item = (&'a Address, Nonce)>,
+    {
+        self.0.with(|c| {
+            let mut msgs = Vec::new();
+            for (sender, mut nonce) in txs {
+                if let Some(buffer) = c.get_mut(sender) {
+                    nonce += 1;
+                    while let Some(msg) = buffer.remove(&nonce) {
+                        msgs.push((*sender, nonce, msg));
+                        nonce += 1;
+                    }
+                }
+            }
+            msgs
+        })
+    }
+}
+
+/// Subscribe to `NewBlock  notifications and clear transactions from the caches.`
+pub fn start_tx_cache_clearing(
+    client: FendermintClient<HybridClient>,
+    tx_cache: TransactionCache,
+    tx_buffer: TransactionBuffer,
+) {
     tokio::task::spawn(async move {
         let chain_id = get_chain_id(&client).await;
-        tx_cache_clearing_loop(client, cache, chain_id).await;
+        tx_cache_clearing_loop(client.into_underlying(), chain_id, tx_cache, tx_buffer).await;
     });
 }
 
-async fn tx_cache_clearing_loop(
-    client: FendermintClient<HybridClient>,
-    tx_cache: TransactionCache,
+async fn tx_cache_clearing_loop<C>(
+    client: C,
     chain_id: ChainID,
-) {
+    tx_cache: TransactionCache,
+    tx_buffer: TransactionBuffer,
+) where
+    C: Client + SubscriptionClient + Send + Sync,
+{
     loop {
         let query = Query::from(EventType::NewBlock);
 
-        match client.underlying().subscribe(query).await {
+        match client.subscribe(query).await {
```

Comment:
If the Client dies, will there be some kind of restart mechanism or resubscription? Then the pruning might be stoped and out of order txns are not submitted.

Reply:
Yes if `.next()` returns an `Err` then we come back here and subscribe again. The subscription will cause a new websocket client to be created.